Split extract MiniSatSolver from MiniSat

cprudhom opened the pull request:
Everything related to search in MiniSat is useless for LCG. So, a subclass can be created only for that purpose, namely MiniSatSolver.

+ activate topLevelCleanUp in LCG

cprudhom committed:
Remove compile as 1st stop of test command in Makefile

cprudhom committed:
Change CustomListener (solver) to reduce log size.
Indeed, since some tests rely on dataproviders and running the tests may produce OOM.

cprudhom committed:
Fix bug introduced when explaining 'abs' for lcg

cprudhom committed:
Rectification of tests following modification of MinCostMaxFlow (bfcf0b26c)

cprudhom committed:
Rectification of tests following correction of BoolEqView (29892370)

cprudhom committed:
Rectification of tests following revamping of reifications (55f60fb0)

cprudhom committed:
Correction of PropAbsolute explanations

jgFages committed:
replace .clear() by .resetQuick() for TIntArrayList

cprudhom committed:
Split extract MiniSatSolver from MiniSat

Everything related to search in MiniSat is useless for LCG.
So, a subclass can be created only for that purpose, namely MiniSatSolver.

+ activate topLevelCleanUp in LCG

jgFages commented on solver/src/main/java/org/chocosolver/sat/MiniSat.java:
```diff
@@ -922,35 +747,45 @@ private int getBacktrackLevel(TIntList out_learnt) {
+        }
+//        cs.shrink(i - j);
+        if (cs.size() > j) {
+            cs.subList(j, cs.size()).clear();
```
rebase branch and check replacement by .resetQuick()

cprudhom replied:
here, `cs.subList` returns a view of cs as a `List` without the `resetQuick()` method.

jgFages commented on solver/src/main/java/org/chocosolver/sat/MiniSatSolver.java:
```diff
@@ -0,0 +1,327 @@
+     */
+    public ESat solve() {
+        model.clear();
+        conflict.clear();
```
rebase branch and check replacement by .resetQuick()

jgFages commented on solver/src/main/java/org/chocosolver/sat/MiniSatSolver.java:
```diff
@@ -0,0 +1,327 @@
+                conflictC++;
+                if (trailMarker() == 0) return ESat.FALSE;
+
+                learnt_clause.clear();
```
rebase branch and check replacement by .resetQuick()

cprudhom committed:
Remove useless assertions in MiniSat (but kept in MiniSatSolver)
Indeed, the assertion tests if a variable is free or in order_heap. But with LCG, no element is popped from order_heap and thus it always contains all variables of the model.

cprudhom committed:
Replace calls to TIntArrayList#clear with calls to #resetQuick (related to #1143)

cprudhom committed:
Rollback unchecked modification on LCG condition to call topLevelCleanUp

cprudhom committed:
Merge branch 'develop' into split-minisat
